File: hyunseok0314/Week_19/BOJ_21278_호석이_두_마리_치킨.py
from itertools import combinations
import sys

input = sys.stdin.readline

N, M = map(int, input().rstrip().split())

dist = [[float("inf") if i != j else 0 for j in range(N)] for i in range(N)]

for _ in range(M):
    u, v = map(lambda x: int(x) - 1, input().rstrip().split())
    dist[u][v] = 1
    dist[v][u] = 1

for k in range(N):
    for i in range(N):
        for j in range(N):
            dist[i][j] = min(dist[i][j], dist[i][k] + dist[k][j])

ans = (float("inf"), float("inf"), float("inf"))

for c in combinations(range(N), 2):
    u, v = c
    ans = min(ans, (sum(map(min, zip(dist[u], dist[v]))) << 1, u + 1, v + 1))

print(ans[1], ans[2], ans[0])

File: sickbirdd/Week_19/BOJ_21278_호석이 두 마리 치킨.java
import java.io.*;
import java.util.*;

public class Main {
	final static int INF = (int) 1e6;
	public static void main(String[] args) throws Exception {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringTokenizer st = new StringTokenizer(br.readLine());
		int N = Integer.parseInt(st.nextToken()), M = Integer.parseInt(st.nextToken());
		int[][] dist = new int[N + 1][N + 1];

		for(int i = 1; i <= N; i++) {
			for(int j = 1; j <= N; j++) {
				dist[i][j] = (i == j) ? 0 : INF;
			}
		}

		for(int i = 0; i < M; i++) {
			st = new StringTokenizer(br.readLine());
			int A = Integer.parseInt(st.nextToken()), B = Integer.parseInt(st.nextToken());
			dist[A][B] = 1;
			dist[B][A] = 1;
		}

		for(int k = 1; k <= N; k++) {
			for(int i = 1; i <= N; i++) {
				for(int j = 1; j <= N; j++) {
					dist[i][j] = Math.min(dist[i][j], dist[i][k] + dist[k][j]);
				}
			}
		}

		int ans = INF, start = -1, end = -1;
		for(int i = 1; i < N; i++) {
			for(int j = i + 1; j <= N; j++) {
				int sum = 0;
				for(int k = 1; k <= N; k++) {
					sum += Math.min(dist[i][k], dist[j][k]);
				}
				if(sum < ans) {
					start = i;
					end = j;
					ans = sum;
				}
			}
		}
		System.out.println(start + " " + end + " " + ans * 2);
	}
}